File: java/cpp/dataStructures/graphs/Flograph.cpp
/** @file Flograph.cpp
 *
 *  @author Jon Turner
 *  @date 2011
 *  This is open source software licensed under the Apache 2.0 license.
 *  See http://www.apache.org/licenses/LICENSE-2.0 for details.
 */
#include "Flograph.h"

#define flo(x) floInfo[x].flo
#define cpy(x) floInfo[x].cpy

/** Constructor for Flograph objects.
 *  @param numv is number of vertices in graph
 *  @param maxe is max possible number of edges in graph
 *  @param s1 is the vertex number of the source
 *  @param t1 is the vertex number of the sink
 */
Flograph::Flograph(int numv, int maxe, int s1, int t1) 
	: Digraph(numv, maxe), s(s1), t(t1) {
        assert(N >= 2 && maxEdge >= 0 && 1 <= s && s <= N 
		&& 1 <= t && t <= N && s != t);
	makeSpace(numv,maxe);
}

/** Copy constructor - not implemented. */
Flograph::Flograph(const Flograph& original) : Digraph(original) { }

Flograph::~Flograph() { freeSpace(); }

/** Allocate and initialize dynamic storage for graph.
 *  @param numv is number of vertices in graph
 *  @param maxe is max possible number of edges in graph
 */
void Flograph::makeSpace(int numv, int maxe) {
        floInfo = new FloInfo[maxe+1];
}

/** Free dynamic storage. */
void Flograph::freeSpace() { delete [] floInfo; }

/** Resize flograph. 
 *  Current graph is discarded, space is freed and new space allocated.
 *  @param numv is number of vertices in graph
 *  @param maxe is max possible number of edges in graph
 */
void Flograph::resize(int numv, int maxe) {
        Digraph::resize(numv, maxe);
        freeSpace(); makeSpace(numv, maxe);
}

void Flograph::reset() { Digraph::reset(); }

/** Copy contents of original to *this.
 *  @param original is a flowgraph which is to be copied to *this
 */
void Flograph::copyFrom(const Flograph& original) {
        if (N != original.n() || maxEdge < original.m()) {
                resize(original.n(), original.m());
        } else {
                reset();
        }
        N = original.n();
        for (edge e = original.first(); e != 0; e = original.next(e)) {
                edge ee = join(original.left(e),original.right(e));
		cpy(ee) = original.cpy(e); flo(ee) = original.flo(e);
        }
	setSrcSnk(original.src(),original.snk());
        sortAdjLists();
}

/** Join two vertices with an edge.
 *  @param u is a vertex number
 *  @param v is a vertex number
 *  @return number of new edge from u to v
 */
edge Flograph::join(vertex u, vertex v) {
	assert(1 <= u && u <= N && 1 <= v && v <= N && M < maxEdge);
	edge e = Digraph::join(u,v); flo(e) = 0;
	return e;
}

/** Remove flow from all edges.
 */
void Flograph::clear() {
	for (edge e = first(); e != 0; e = next(e))
		addFlow(tail(e),e,-f(tail(e),e));
}

/** Add to the flow on an edge.
 *  @param v is a vertex
 *  @param e is an edge with v as one of its endpoints
 *  @param ff is a flow to be added to e, from v to the other endpoint
 *  @return the amount of flow on the edge leaving v, after ff has been added
 */
flow Flograph::addFlow(vertex v, edge e, flow ff) {
	if (tail(e) == v) {
		if (ff + flo(e) > cpy(e) || ff + flo(e) < 0) {
			fatal("addflow: requested flow outside allowed range");
		}
		flo(e) += ff;
		return flo(e);
	} else {
		if (flo(e) - ff < 0 || flo(e) - ff > cpy(e)) {
			fatal("addflow: requested flow outside allowed range");
		}
		flo(e) -= ff;
		return -flo(e);
	}
}

/** Read an edge from the input stream and add it to the flograph.
 *  @param in is an open input stream
 *  @return true if the operation is successful, else false
 */
bool Flograph::readEdge(istream& in) {
        vertex u, v; flow capp, ff;
        if (Util::readNext(in,'(') == 0 || !Util::readNode(in,u,N) ||
            Util::readNext(in,',') == 0 || !Util::readNode(in,v,N) ||
            Util::readNext(in,',') == 0 || !Util::readNum(in,capp) ||
            Util::readNext(in,',') == 0 || !Util::readNum(in,ff) || 
	    Util::readNext(in,')') == 0)
		return false;

	if (u < 1 || u > n() || v < 1 || v > n()) return false;
        edge e; e = join(u,v);
	if (e == 0) return false;
	setCapacity(e,capp); addFlow(u,e,ff);

        return true;

}

/** Read a flograph from an input stream and assign to this object.
 *  @param in is an open input stream
 *  @return true on success, false on failure
 */
bool Flograph::read(istream& in) {
        int numv, nume;
        in >> numv >> nume;
        if (numv != N || nume > maxEdge) resize(numv,nume);
	else reset();
        N = numv;

	vertex src, snk;
	if (!Util::readNode(in,src,N) || !Util::readNode(in,snk,N))
		return false;
	setSrcSnk(src,snk);

	for (int i = 1; i <= nume; i++)
		if (!readEdge(in)) return false;
	if (M != nume) return false;
        return true;
}

/** Create readable representation of an edge.
 *  @param e is an edge
 *  @param s is a string in which result is to be returned
 *  @return a reference to s
 */
string& Flograph::edge2string(edge e, string& s) const {
	s = "";
	vertex u = tail(e); vertex v = head(e);
        if (e == 0) {
               s += "-"; 
	} else {
		string s1;
		s += "(" + Util::node2string(u,N,s1); 
		s += "," + Util::node2string(v,N,s1); 
		s += "," + Util::num2string(cap(u,e),s1);
		s += "," + Util::num2string(f(u,e),s1) + ")";
        }
	return s;
}

/** Create readable representation of this flograph.
 *  @param s is a string in which result is to be returned
 *  @return a reference to s
 */
string& Flograph::toString(string& s) const {
	int i; vertex u; edge e;
	stringstream ss;
	ss << n() << " " << m() << " ";
	ss << Util::node2string(src(),n(),s) << " ";
	ss << Util::node2string(snk(),n(),s) << "\n";
	for (u = 1; u <= n(); u++) {
		i = 0;
		for (e = firstOut(u); e != 0; e = nextOut(u,e)) {
			ss << edge2string(e,s) << " ";
                        if ((++i)%5 == 0) ss << "\n";
                }
                if (i>0 && i%5 != 0) ss << "\n";
	}
	s = ss.str();
	return s;
}

/** Create graphviz representation of this flograph.
 *  @param s is a string in which result is to be returned
 *  @return a reference to s
 */
string& Flograph::toDotString(string& s) const {
	int i; vertex u; edge e;
	stringstream ss;
	ss << "digraph G { " << endl;
        ss << Util::node2string(src(),n(),s) 
           << " [ style = bold, peripheries = 2, color = red]; "
           << endl;
	ss << Util::node2string(snk(),n(),s) 
           << " [ style = bold, peripheries = 2, color = blue]; "
           << endl;
	for (u = 1; u <= n(); u++) {
                string su;
		for (e = firstOut(u); e != 0; e = nextOut(u,e)) {
                        vertex u = tail(e); vertex v = head(e);
                        string s1;
                        if (e != 0) {
                                su += Util::node2string(u,N,s1) + " -> "; 
                                s1.clear();
                                su += Util::node2string(v,N,s1); 
                                s1.clear(); su += " [label = \" ";
                                su += "(" + Util::num2string(cap(u,e),s1); 
                                s1.clear();
                                su += "," + Util::num2string(f(u,e),s1) 
                                        + ") \"]; "; 
                        }
                }
                if (!su.empty())   ss << su << endl;
	}
	ss << " } " << endl;
	s = ss.str();
	return s;
}

/** Shuffle the vertices and edges according to given permutations.
 *  @param vp is a permutation on the integers 1..(# of vertices)
 *  @param ep is a permutation on the integers 1..(# of edges)
 *  This method remaps vertex u to vp[u] and edge e to ep[e]
 */
void Flograph::shuffle(int vp[], int ep[]) {
        edge e;
	FloInfo *floInfo1 = new FloInfo[M+1];

        Digraph::shuffle(vp,ep);
        for (e = 1; e <= M; e++) floInfo1[ep[e]] = floInfo[e];
        for (e = 1; e <= M; e++) floInfo[e] = floInfo1[e];
	s = vp[s]; t = vp[t];

        delete [] floInfo1;
}

/** Generate random flow graph.
 *  @param numv is the number of vertices in the generated graph
 *  @param nume is the number of edges in the generated graph
 *  @param mss is the out-degree of the source and the in-degree of the sink
 *  
 *  The generated graph has a "core" subgraph with numv-2 nodes and
 *  nume-2*mss edges and the specified span. It is generated using
 *  Digraph::rgraph. The source and sink are then added to the core.
 *  If mss>(numv-2)/4, it is first reduced to (numv-2)/4.
 */
void Flograph::rgraph(int numv, int nume, int mss) {
	mss = max(1,mss); mss = min(mss,(numv-2)/4); 
	numv = max(numv,3); nume = max(2*mss,nume);

	if (N != numv || maxEdge < nume) resize(numv,nume); 
	else reset();

	Digraph::rgraph(numv-2,nume-2*mss);
	N = numv;
	setSrcSnk(numv-1,numv);

	vertex *neighbors = new vertex[2*mss+1];
	Util::genPerm(2*mss,neighbors);
	for (int i = 1; i <= mss; i++) {
		join(src(),neighbors[i]);
	}
	Util::genPerm(2*mss,neighbors);
	for (int i = 1; i <= mss; i++) {
		join(((numv-2)/2)+neighbors[i],snk());
	}
	sortAdjLists();

	delete [] neighbors;
}

/** Generate random edge capacities.
 *  @param ec1 is the max capacity of the source/sink edges
 *  @param ec2 is the max capacity of the remaining edges
 *  edge capacities are selected uniformly with a minimum of 1
 */
void Flograph::randCapacity(flow ec1, flow ec2) {
	for (edge e = first(); e != 0; e = next(e)) {
		if (tail(e) == s || head(e) == t)
			setCapacity(e,randint(1,ec1));
		else
			setCapacity(e,randint(1,ec2));
	}
}

File: js/dataStructures/graphs/Flograph.mjs
/** @file Flograph.mjs
 *
 *  @author Jon Turner
 *  @date 2021
 *  This is open source software licensed under the Apache 2.0 license.
 *  See http://www.apache.org/licenses/LICENSE-2.0 for details.
 */

import List from '../basic/List.mjs';
import Scanner from '../basic/Scanner.mjs';
import Digraph from './Digraph.mjs';
import { shuffle } from '../../common/Random.mjs';

import { assert, assertEnabled } from '../../common/Assert.mjs';
let ae; // initialized in constructor

/** Data structure for Flograph, used by max flow algorithms.
 *  Extends Digraph class and adds edge capacities, flows and
 *  optionally, minimum flow requirements and costs.
 */
export default class Flograph extends Digraph {
	Source;        // source vertex
	Sink;          // sink vertex

	ssCapScale;     // scaling factor used by randomCapacities() method
					// is set by RandomGraph.randomFlograph() method

	/** Constructor for flow graph
	 *  @param n is the number of vertices
	 *  @param erange is the max number of edges to provide space for
	 */
	constructor(n, erange) {
	ae = assertEnabled();
		super(n, erange); 
		this.Source = 1; this.Sink = this.n;

		this.addEdgeProperty('cap', 1);
		this.addEdgeProperty('flow', 0);

		this.ssCapScale = 1;
	} 

	/** Get/set the source vertex.
	 *  @param s is the new source vertex.
	 *  @return the source
	 */
	get source() { return this.Source; }
	set source(s) { this.Source = s; }

	/** Get/set the sink vertex.
	 *  @param t is the new sink vertex
	 *  @return the sink
	 */
	get sink() { return this.Sink; }
	set sink(t) { this.Sink = t; }

	/** Get the residual capacity of an edge.
	 *  @param u is a vertex in the flograph
	 *  @param e is an edge that is incident to u
	 *  @return the unused capacity of e, going from u to mate(u)
	 */
	res(e, u=this.tail(e)) {
		return (u == this.tail(e) ?
					this.cap(e) - this.flow(e) :
					this.flow(e) - (this.floor ? this.floor(e) : 0));
	}

	/** Get the flow on an edge relative to one of its endpoints.
	 *  @param e is an edge
	 *  @param u is an endpoint of e
	 *  @return the flow on e, going from u to mate(u)
	 */
	f(e, u=this.tail(e)) {
		return (u == this.tail(e) ? this.flow(e) : -this.flow(e));
	}

	/** Get the cost of an edge relative to one of its endpoints.
	 *  @param e is an edge
	 *  @param u is an endpoint of e
	 *  @return the cost of the flow on e, going from u to mate(u)
	 */
	c(e, u=this.tail(e)) {
		return (u == this.tail(e) ? this.cost(e) : -this.cost(e));
	}

	/** Add flow to an edge.
	 *  @param e is an edge
	 *  @param u is an endpoint of e
	 *  @param f is a flow amount to be added to the flow on e leaving u
	 */
	addFlow(e, u, f) {
		ae && assert(f <= this.res(e, u), 'addFlow: edge capacity violation');
		this.flow(e, this.flow(e) + (u == this.tail(e) ? f : (-f)));
	}

	/** Set the flow of every edge to zero. */
	clearFlow() {
		for (let e = this.first(); e; e = this.next(e))
			this.flow(e, 0);
	}

	/** Return the current flow magnitude (total flow leaving source). */
	totalFlow() {
		let flow = 0; let src = this.source;
		for (let e = this.firstOutof(src); e; e = this.nextOutof(src, e)) {
			flow += this.f(e);
		}
		return flow;
	}

	/** Return the set of vertices reachable from the source
	 *  with respect to the current flow.
	 */
	reachable() {
		let q = new List(this.n); let reached = new List(this.n);
		q.enq(1); reached.enq(1);
		while (!q.empty()) {
			let u = q.deq();
			for (let e = this.firstAt(u); e; e = this.nextAt(u,e)) {
				let v = this.mate(u,e);
				if (!reached.contains(v) && !q.contains(v) &&
					this.res(e,u) > 0) {
					q.enq(v); reached.enq(v);
				}
			}
		}
		return reached;
	}

	/** Return flow statistics. 
	 *  @return an object containing several statistics about the current
	 *  flow (assumed to be a max flow); these include the value of the
	 *  flow, the number of vertices reachable from the source (not counting
	 *  source), the number of edges crossing the cut and the capacity of
	 *  the edges crossing the cut.
	 */
	flowStats() {
		let f = this.totalFlow();
		let reachable = this.reachable();
		let cutsize = 0; let cutcap = 0;
		for (let u = reachable.first(); u != 0; u = reachable.next(u)) {
			for (let e = this.firstOutof(u); e; e = this.nextOutof(u, e)) {
				if (reachable.contains(this.head(e))) continue;
				cutsize++; cutcap += this.cap(e);
			}
		}
		return { 'totalFlow': f, 'reachable': reachable.length-1,
				 'cutsize': cutsize, 'cutcap': cutcap };
	}

	/** Return the current flow cost (sum of flow*cost for all edges). */
	totalCost() {
		let cost = 0;
		for (let e = this.first(); e; e = this.next(e)) {
			cost += this.flow(e) * this.cost(e);
		}
		return cost;
	}

	/** Join two vertices with an edge.
	 *  @param u is a vertex
	 *  @param v is an edge
	 *  @param e is a currently unused edge number to be used for the
	 *  new edge (u,v); if omitted, the first available edge number is used
	 *  the edge capacity and flow are both initialized to zero,
	 *  as is the min flow requirement, if enabled
	 */
	join(u, v, e=this.edges.first(2)) {
		let ee = super.join(u, v, e);
//console.log('fgjoin', e, ee, this.edgeRange, this.validEdge(ee), this.edges.in(ee,1));
		this.cap(ee, 0); this.flow(ee, 0);
		if (this.cost) this.cost(ee, 0);
		if (this.floor) this.floor(ee, 0);
		return ee;
	}

	/** Compare another flograph to this one.
	 *  @param that is a Flograph object or a string representing one.
	 *  @param includeFlow (when true) causes the comparison to include
	 *  flows on edges
	 *  @return true if that is equal to this; that is, it has the same
	 *  vertices, edges, capacities, costs, floors (and possibly flows);
	 *  endpoint lists are sorted as a side effect
	 */
	equals(that, includeFlow=false) {
		// note: not leveraging superclass due to conflicting requirements
		if (this === that) return true;
        if (typeof that == 'string') {
            let s = that;
			that = Flograph.fromString(s, this.n, this.edgeRange);
			assert(that != null, 
				   'Flograph.fromString: called by .equals() cannot parse ' +s);
        } else if (that.constructor.name != 'Flograph') {
			return false;
		}
		if (!!this.cost  != !!that.cost  || !!this.floor != !!that.floor)
			return false;
		let el1 = this.sortedElist(); let el2 = that.sortedElist();
		for (let i = 0; i < el1.length; i++) {
			let e1 = el1[i]; let e2 = el2[i];
			if (this.cap(e1) != that.cap(e2) ||
				(includeFlow && this.flow(e1) != that.flow(e2)) ||
				this.cost  && this.cost(e1)  != that.cost(e2)   ||
				this.floor && this.floor(e1) != that.floor(e2)    ) {
				return false;
			}
		}
		return that;
	}

	/** Compare two edges incident to the same endpoint u.
	 *  @return -1 if u's mate in e1 is less than u's mate in e2,
	 *  return +1 if u's mate in e1 is greater than than u's mate in e2,
	 *  return  0 if u's mate in e1 is equal to its mate in e2.
	 */
	ecmp(e1, e2, u) {
		ae && assert(this.validVertex(u) && this.validEdge(e1)
										 && this.validEdge(e2));
			 if (u == this.head(e1) && u == this.tail(e2)) return -1;
		else if (u == this.tail(e1) && u == this.head(e2)) return 1;

		let v1 = this.mate(u, e1); let v2 = this.mate(u, e2);
		let cap1 = this.cap(e1); let cap2 = this.cap(e2);

		return (v1 != v2 ? v1 - v2 :
				(cap1 != cap2 ? cap1 - cap2 :
				 (this.cost && this.cost(e1) != this.cost(e2) ?
							   this.cost(e1)  - this.cost(e2) :
				  (this.floor && this.floor(e1) != this.floor(e2) ?
								 this.floor(e1)  - this.floor(e2) : 0))));
	}

	/** Create a string representation of an edge.
	 *  @param e is an edge number
	 *  @return a string representing the edge
	 */
	edge2string(e, label) {
		if (e == 0) return '-';
		return '(' + this.x2s(this.tail(e), label) + ',' 
				   + this.x2s(this.head(e), label) + ','
				   + (this.floor && this.floor(e)>0 ? this.floor(e) + '-' : '')
				   + this.cap(e)
				   + (this.cost && this.cost(e) ? '@' + this.cost(e) : '') 
				   + (this.f(e)>0 ? '/' + this.f(e) : '') + ')';
	}

	/** Construct a string representation of the Flograph object.
	 *  @param fmt is an integer; its low order bits specify format options
	 *		0b0001 causes the adjacency lists to be shown on separate lines
	 *		0b0010 causes empty lists to be shown explicitly
	 *		0b0100 omits edges from the list of the "larger" endpoint
	 *      0b1000 omits edges with zero flow
	 *  Uses elab and vlab  to modify the behavior of the Graph.toString()
	 *  method.
	 */
	toString(fmt=2, elab=0, vlab=0) {
		if (!elab) {
			elab = (e,u) => 
					(u == this.head(e) || (fmt&0x8 && this.flow(e) == 0) ? '' :
					 (this.x2s(this.mate(u,e)) + ':' +
					  (this.floor && this.floor(e)>0 ?
						this.floor(e) + '-' : '') +
				   	  this.cap(e) +
					  (this.cost && this.cost(e) ? '@' + this.cost(e) : '') +
					  (this.flow(e)>0 ? '/' + this.flow(e) : '')
					 ));
		}
		if (!vlab) {
			vlab = (u) =>
						(u == this.source ? this.x2s(u) + '->' :
						 (u == this.sink ? '->' + this.x2s(u) : this.x2s(u)));
		}
		return super.toString(fmt, elab, vlab);
	}
		
	/** Initialize graph from a string representation.
	 *  @param s is a string representing a graph
	 *  @return true on success, else false
	 */
	static fromString(s, n=10, erange=10) {
		let source = 0; let sink = 0;
		// function to parse a vertex
		let vnext = (sc => {
						let u;
						if (sc.verify('->')) {
							u = sc.nextIndex();
							if (sink) return false;
							sink = u;
						} else {
							u = sc.nextIndex();
							if (sc.verify('->')) {
								if (source || source == u) return false;
								source = u;
							}
						}
						n = Math.max(n,u);
						return u > 0 ? u : 0;
					});
		// function to parse an adjacency list item and save properties
		let pairs=[]; let floors=[]; let caps=[]; let flows=[]; let costs=[];
		let hasFloors = 0; let hasCosts = 0;
		let enext = ((u,sc) => {
						let v = sc.nextIndex();
						if (v < 0) return false;
						n = Math.max(n,v);
						let i = pairs.length;
						pairs.push([u,v]);
						if (!sc.verify(':',0)) return false;
						let x = sc.nextNumber();
						if (Number.isNaN(x)) return false;
						if (!sc.verify('-')) {
							caps[i] = x;
						} else {
							floors[i] = x;
							hasFloors = 1;
							let cap = sc.nextNumber();
							if (Number.isNaN(cap)) return false; 
							caps[i] = cap;
						}
						if (sc.verify('@')) {
							let cost = sc.nextNumber();
							if (Number.isNaN(cost)) return false;
							costs[i] = cost;
							hasCosts = 1;
						}
						if (sc.verify('/')) {
							let flow = sc.nextNumber();
							if (Number.isNaN(flow)) return false;
							flows[i] = flow;
						}
						return true;
					});

		let sc = new Scanner(s);
		if (!sc.verify('{')) return null;
		while (!sc.verify('}')) {
			let u = vnext(sc);
			if (!u) return null;
			if (sc.verify('[')) {
				while (!sc.verify(']')) {
					if (!enext(u,sc)) return null;
				}
			}
		}

		let g = new Flograph(n, Math.max(erange, pairs.length));
		if (hasCosts) g.addEdgeProperty('cost', 0);
		if (hasFloors) g.addEdgeProperty('floor', 0);

		// configure graph
		for (let i = 0; i < pairs.length; i++) {
			let [u,v] = pairs[i];
			let e = g.join(u,v);
			g.cap(e, caps[i]);
			if (flows[i]) g.flow(e, flows[i]);
			if (g.cost && costs[i]) g.cost(e, costs[i]);
			if (g.floor && floors[i]) g.floor(e, floors[i]);
		}
		g.source = source; g.sink = sink;
		return g;
	}

	/** Randomize the order of the vertices and edges. */
	scramble() {
		let ep = super.scramble();
		shuffle(this.F, ep); shuffle(this.Cap, ep);
	}

	/** Compute random capacities for edges.
	 *  @param rand is a random number generator used to generate the
	 *  random edge capacities; it is invoked using any extra arguments
	 *  provided by caller; for example randomCapacities(randomInteger, 1, 10)
	 *  will assign random integer capacities in 1..10.
	*/
	randomCapacities(rand, ...args) {
		let s = this.source; let t = this.sink;
		for (let e = this.first(); e; e = this.next(e)) {
			let c = rand(...args);
			if (this.tail(e) == s || this.head(e) == t) c *= this.ssCapScale;
			this.cap(e, Math.max(c, (this.floor ? this.floor(e) : 0)));
		}
	}

	/** Compute random costs for all the edges.
	 *  @param rand is a random number generator used to generate the
	 *  random edge costs; it is invoked using any extra arguments
	 *  provided by caller; for example randomCosts(randomInteger, 1, 10)
	 *  will assign random integer costs in 1..10.
	 */
	randomCosts(rand, ...args) {
		if (!this.cost) this.addEdgeProperty('cost', 0);
		for (let e = this.first(); e; e = this.next(e)) {
			this.cost(e, rand(...args));
		}
	}

	/** Compute random floor values for all the edges.
	 *  @param rand is a random number generator used to generate the
	 *  random edge costs; it is invoked using any extra arguments
	 *  provided by caller; for example randomFloors(randomInteger, 1, 10)
	 *  will assign random integer costs in 1..10.
	 */
	randomFloors(rand, ...args) {
		if (!this.floor) this.addEdgeProperty('floor', 0);
		for (let e = this.first(); e; e = this.next(e)) {
			let f = rand(...args);
			this.floor(e, Math.min(f, this.cap(e)));
		}
	}
}
